src/main/java/com/unascribed/fabrication/mixin/i_woina/end_portal_parallax/MixinEndPortalRenderer.java

```diff
@@ -0,0 +1,178 @@
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+
+import java.nio.FloatBuffer;
+import java.nio.IntBuffer;
+import java.util.Random;
+
+@Mixin(EndPortalBlockEntityRenderer.class)
+@EligibleIf(configAvailable="*.end_portal_parallax", envMatches=Env.CLIENT, specialConditions=SpecialEligibility.NOT_MACOS)
+public abstract class MixinEndPortalRenderer {
+
+	@Shadow
+	@Final
+	public static Identifier PORTAL_TEXTURE;
+	@Shadow
+	@Final
+	public static Identifier SKY_TEXTURE;
+	@Unique
+	private static final Random fabrication$random = new Random(31100L);
+	private static final FloatBuffer fabrication$matrix_buffer = BufferUtils.createFloatBuffer(4*4);
+
+	@FabInject(at=@At("HEAD"), method= "renderSide(Lnet/minecraft/block/entity/EndPortalBlockEntity;Lnet/minecraft/util/math/Matrix4f;Lnet/minecraft/client/render/VertexConsumer;FFFFFFFFLnet/minecraft/util/math/Direction;)V", cancellable=true)
+	public void oldRender(EndPortalBlockEntity be, Matrix4f model, VertexConsumer vertices, float x1, float x2, float y1, float y2, float z1, float z2, float z3, float z4, Direction side, CallbackInfo ci) {
+		if (!FabConf.isEnabled("*.end_portal_parallax")) return;
+		ci.cancel();
+		if (!be.shouldDrawSide(side)) return;
+		MinecraftClient mc = MinecraftClient.getInstance();
+		try (MemoryStack stack = MemoryStack.stackPush()) {
+			IntBuffer viewportBuffer = stack.mallocInt(4);
+			FloatBuffer scratchBuffer = stack.mallocFloat(16);
+			FloatBuffer objectCoords = stack.mallocFloat(3);
+			FloatBuffer projBuffer = stack.mallocFloat(16);
+
+			GL11.glGetIntegerv(GL11.GL_VIEWPORT, viewportBuffer);
+			Matrix4f scratchMat = RenderSystem.getModelViewMatrix().copy();
+			Matrix4f projMatrix = RenderSystem.getProjectionMatrix().copy();
+			scratchMat.multiply(model);
+			projMatrix.writeRowMajor(projBuffer);
+			scratchMat.writeRowMajor(scratchBuffer);
+
+			float objectX = objectCoords.get(0);
+			float objectY = objectCoords.get(1);
+			float objectZ = objectCoords.get(2);
+
+			BlockPos pos = be.getPos();
+			float baseTexTransX = pos.getX();
+			float baseTexTransY = pos.getY();
+			float baseTexTransZ = pos.getZ();
+
+			GL11.glMatrixMode(GL11.GL_PROJECTION);
+			GL11.glPushMatrix();
+			GL11.glLoadIdentity();
+			RenderSystem.getProjectionMatrix().writeColumnMajor(fabrication$matrix_buffer);
+			fabrication$matrix_buffer.rewind();
+			GL11.glMultMatrixf(fabrication$matrix_buffer);
+			GL11.glMatrixMode(GL11.GL_MODELVIEW);
+			GL11.glPushMatrix();
+			RenderSystem.getModelViewMatrix().writeColumnMajor(fabrication$matrix_buffer);
+			fabrication$matrix_buffer.rewind();
+			GL11.glMultMatrixf(fabrication$matrix_buffer);
+			model.writeColumnMajor(fabrication$matrix_buffer);
+			fabrication$matrix_buffer.rewind();
+			GL11.glMultMatrixf(fabrication$matrix_buffer);
+			GL11.glEnable(GL11.GL_TEXTURE_2D);
+			GL11.glEnable(GL11.GL_DEPTH_TEST);
+			GL11.glDepthFunc(GL11.GL_LEQUAL);
+			GL11.glDisable(GL11.GL_LIGHTING);
```

Comment:
I had requested depending on Lib39 Deferral instead of just copy-pasting everything, but whatever I guess. Please `import static` GL11.